File: DependencyInjection/ChangingBehaviorThroughComposition/CompositionRoot.cs
namespace ChangingBehaviorThroughComposition
{
  public class CompositionRoot
  {
    private const string SecurityPhoneNumber = "11-222-1121";
    private const string PolicePhoneNumber = "122-33-22-12";

    #region hide
    public void ALoudAlarm()
    {
      var building = new Building(new LoudAlarm());
    }
    #endregion

    #region hide
    public void ASilentAlarm()
    {
      var building = new Building(new SilentAlarm(PolicePhoneNumber));
    }
    #endregion

    #region hide
    public void AlarmDisabled()
    {
      var building = new Building(new NoAlarm());
    }
    #endregion

    #region hide
    public void TwoSilentAlarms()
    {
      var building = new Building(
        new CompoundAlarm(
          new SilentAlarm(PolicePhoneNumber),
          new SilentAlarm(SecurityPhoneNumber)));
    }
    #endregion

    #region hide
    public void TwoSilentAlarmsAndOneLoud()
    {
      var building = new Building(
        new CompoundAlarm(
          new SilentAlarm(PolicePhoneNumber),
          new SilentAlarm(SecurityPhoneNumber),
          new LoudAlarm()));
    }
    #endregion

    #region hide
    public void TwoSilentAlarmsAndOneLoudAtNightAndNoneDuringDay()
    {
      var building = new Building(
        new SwitchableAlarm(new MetDuringTheDayCriteria(), 
          new NoAlarm(),
          new CompoundAlarm(
            new SilentAlarm(PolicePhoneNumber),
            new SilentAlarm(SecurityPhoneNumber),
            new LoudAlarm())
          ));
    }
    #endregion

    #region hide
    public void ByDefaultNoAlarmButTwoSilentAlarmsAndOneLoudOnlyAtNightWeekend()
    {
      var building = new Building(
        new SwitchableAlarm(
          new CompoundSwitchCriteria(
            new MetDuringTheDayCriteria(), 
            new MetOnWorkDayCriteria()),
          new NoAlarm(),
          new CompoundAlarm(
            new SilentAlarm(PolicePhoneNumber),
            new SilentAlarm(SecurityPhoneNumber),
            new LoudAlarm())
          ));
    }
    #endregion

    #region hide
    public void ByDefaultNoAlarmButTwoSilentAlarmsAndOneLoudOnlyAtNightWeekend__InitializedInParallel()
    {
      var building = new Building(
        new SwitchableAlarm(
          new CompoundSwitchCriteria(
            new MetDuringTheDayCriteria(), 
            new MetOnWorkDayCriteria()),
          new NoAlarm(),
          new ParallelCompoundAlarm( //!!!
            new SilentAlarm(PolicePhoneNumber),
            new SilentAlarm(SecurityPhoneNumber),
            new LoudAlarm())
          ));
    }
    #endregion

  
  /* Conclusions:
   * 1. Changing behavior by composition
   * 2. Did not have to modify any code but composition
   * 3. Separate use from construction - SRP
   */

    #region hide - DSL - show later!
    public void ByDefaultNoAlarmButTwoSilentAlarmsAndOneLoudOnlyAtNightWeekend__InitializedInParallel_FluentInterface()
    {
      var building = new Building(
        Switch(
          IfBoth(IsDay(), IsNotWeekend()), 
            DoNothing(),
          Else(
            Call(PolicePhoneNumber),
            Call(SecurityPhoneNumber),
            ActivateSirens())));
    }

    private static NoAlarm DoNothing()
    {
      return new NoAlarm();
    }

    private static LoudAlarm ActivateSirens()
    {
      return new LoudAlarm();
    }

    private Alarm Call(string policePhoneNumber)
    {
      return new SilentAlarm(policePhoneNumber);
    }

    private Alarm Else(params Alarm[] alarms)
    {
      return new ParallelCompoundAlarm(alarms);
    }

    private static MetOnWorkDayCriteria IsNotWeekend()
    {
      return new MetOnWorkDayCriteria();
    }

    private static MetDuringTheDayCriteria IsDay()
    {
      return new MetDuringTheDayCriteria();
    }

    private SwitchCriteria IfBoth(params SwitchCriteria[] criterias)
    {
      return new CompoundSwitchCriteria(criterias);
    }


    public SwitchableAlarm Switch(SwitchCriteria switchCriteria, Alarm defaultAlarm, Alarm secondaryAlarm)
    {
      return new SwitchableAlarm(switchCriteria, defaultAlarm, secondaryAlarm);
    }


    #endregion
  }
}

File: Java/dependency-injection/changing-behavior-through-composition/src/main/java/com/github/grzesiek_galezowski/ChangingBehaviorThroughComposition/CompositionRoot.java
package com.github.grzesiek_galezowski.ChangingBehaviorThroughComposition;

import com.github.grzesiek_galezowski.ChangingBehaviorThroughComposition.Alarms.*;
import com.github.grzesiek_galezowski.ChangingBehaviorThroughComposition.SwitchCriterias.CompoundSwitchCriteria;
import com.github.grzesiek_galezowski.ChangingBehaviorThroughComposition.SwitchCriterias.MetDuringTheDayCriteria;
import com.github.grzesiek_galezowski.ChangingBehaviorThroughComposition.SwitchCriterias.MetOnWorkDayCriteria;

public class CompositionRoot {
  private static final String SecurityPhoneNumber = "11-222-1121";
  private static final String PolicePhoneNumber = "122-33-22-12";

  public void aLoudAlarm() {
    Building building = new Building(new LoudAlarm());
  }

  public void aSilentAlarm() {
    Building building = new Building(new SilentAlarm(PolicePhoneNumber));
  }

  public void alarmDisabled() {
    Building building = new Building(new NoAlarm());
  }

  public void twoSilentAlarms() {
    Building building = new Building(
        new CompoundAlarm(
            new SilentAlarm(PolicePhoneNumber),
            new SilentAlarm(SecurityPhoneNumber)));
  }

  public void twoSilentAlarmsAndOneLoud() {
    Building building = new Building(
        new CompoundAlarm(
            new SilentAlarm(PolicePhoneNumber),
            new SilentAlarm(SecurityPhoneNumber),
            new LoudAlarm()));
  }

  public void twoSilentAlarmsAndOneLoudAtNightAndNoneDuringDay() {
    Building building = new Building(
        new SwitchableAlarm(new MetDuringTheDayCriteria(),
            new NoAlarm(),
            new CompoundAlarm(
                new SilentAlarm(PolicePhoneNumber),
                new SilentAlarm(SecurityPhoneNumber),
                new LoudAlarm())
        ));
  }

  public void byDefaultNoAlarmButTwoSilentAlarmsAndOneLoudOnlyAtNightWeekend() {
    Building building = new Building(
        new SwitchableAlarm(
            new CompoundSwitchCriteria(
                new MetDuringTheDayCriteria(),
                new MetOnWorkDayCriteria()),
            new NoAlarm(),
            new CompoundAlarm(
                new SilentAlarm(PolicePhoneNumber),
                new SilentAlarm(SecurityPhoneNumber),
                new LoudAlarm())
        ));
  }

  public void byDefaultNoAlarmButTwoSilentAlarmsAndOneLoudOnlyAtNightWeekendInitializedInParallel() {
    Building building = new Building(
        new SwitchableAlarm(
            new CompoundSwitchCriteria(
                new MetDuringTheDayCriteria(),
                new MetOnWorkDayCriteria()),
            new NoAlarm(),
            new ParallelCompoundAlarm( //!!!
                new SilentAlarm(PolicePhoneNumber),
                new SilentAlarm(SecurityPhoneNumber),
                new LoudAlarm())
        ));
  }

  /* Conclusions:
   * 1. Changing behavior by composition
   * 2. Did not have to modify any code but composition
   * 3. Separate use from construction - SRP
   */

  public void byDefaultNoAlarmButTwoSilentAlarmsAndOneLoudOnlyAtNightWeekendInitializedInParallelFluentInterface() {
    Building building = new Building(
        switchBetween(
            doNothing(), ifBoth(isDay(), isNotWeekend()),
            otherwise(
                call(PolicePhoneNumber),
                call(SecurityPhoneNumber),
                activateSirens())));
  }

  private static NoAlarm doNothing() {
    return new NoAlarm();
  }

  private static LoudAlarm activateSirens() {
    return new LoudAlarm();
  }

  private Alarm call(String policePhoneNumber) {
    return new SilentAlarm(policePhoneNumber);
  }

  private Alarm otherwise(Alarm... alarms) {
    return new ParallelCompoundAlarm(alarms);
  }

  private static MetOnWorkDayCriteria isNotWeekend() {
    return new MetOnWorkDayCriteria();
  }

  private static MetDuringTheDayCriteria isDay() {
    return new MetDuringTheDayCriteria();
  }

  private SwitchCriteria ifBoth(SwitchCriteria ...criterias) {
    return new CompoundSwitchCriteria(criterias);
  }

  public SwitchableAlarm switchBetween(Alarm defaultAlarm, SwitchCriteria switchCriteria, Alarm secondaryAlarm) {
    return new SwitchableAlarm(switchCriteria, defaultAlarm, secondaryAlarm);
  }
}
